IPNI_MH_SAMPLING.md

```diff
@@ -0,0 +1,274 @@
+include:
+
+* **Deterministic Sampling Algorithm**: The API employs a predefined algorithm that yields identical sampling results
+  for the same input parameters across a node, even as new multihashes are incrementally ingested within the same
+  context ID by the same provider.
+
+* **Federation Epoch Awareness**: The API maintains consistent multihash sampling across indexer instances within the
+  same federation by considering the federation epoch.
+
+* **Seed-Based Sampling**: Users can specify a seed derived from decentralized randomness beacons like Drand. This
+  guarantees deterministic sampling and just-in-time candidate selection. Queries repeated with the same seed produce
+  identical results, facilitating reproducibility.
+
+The sampling process utilizes the PCG (Permuted Congruential Generator) random number generator, seeded to ensure
+deterministic output. To efficiently perform sampling, two pre-emptive data organizational strategies are proposed:
+
+1. **Counting Multihashes per Context ID**: Indexers pre-calculate the total multihash count per provider ID for each
+   context ID.
+2. **Sorting Multihashes by Hash Value**: Indexers store multihash records grouped by provider ID and context ID, sorted
+   by hash value. This complements a global multihash index linked to provider information.
+
+The sampling process involves initializing the PCG with the seed and selecting the required indices with the upper bound
+being the total multihashes per context ID. The selected multihashes are then retrieved from the sorted list by hash
+value.
+
+While other deterministic sampling algorithms don't require these pre-emptive data organizational strategies, the
+proposed method is more efficient for large datasets. It also supports efficient bulk deletions and deeper meta-analysis
+of indexed records, such as:
+
+* The number of records advertised by a provider per context ID.
+* The number of unique content IDs advertised by a provider.
+* The total number of unique hashes known by an indexer.
+* And more.
+
+A notable benefit of this approach is efficient bulk deletion. The sorted storage of multihashes allows for more
+efficient deletion of ranges rather than individual multihashes. This capability reduces the need for opportunistic
+deletions during lookup flows, which have low cleanup rates and high write amplification.
+
+### API
+
+#### `GET /ipni/v0/sample/{provider-id}/{context-id}`
+
+Samples a set of multihashes ingested by an IPNI indexer for a given provider ID and Context ID.
+
+- **Path Parameters**:
+    - `provider-id`: (string, required) The multibase encoded peer ID of the provider.
+        - Example: `12D3KooWN34sqTaMfZE3ReELyVF3no3qU7883Mi6j2VWsv6dwhPL`
+    - `context-id`: (string, required) The multibase encoded context ID.
+
+- **Query Parameters**:
+    - `seed`: (optional) The seed for deterministic sampling. Ensures repeatability of samples.
```

Comment:
In my limited experience with RNGs, the seed is usually a number. 

However, the DRAND beacon is a string, e.g. `3439d92d58e47d342131d446a3abe264396dd264717897af30525c98408c834f`. I believe this string is a hex-encoded representation of some binary data (a hash digest 32-bytes long?) so we can easily convert it into a big number - 23622477310538551577113686196064208635296596430983311976502637696987544158208. Would your implementation support such big numbers?

As a developer integrating with this new endpoint, I don't know what kind of seed values the API expects and supports.

Let's make the spec a bit more detailed, please, and describe the seed value type plus any constraints like maximum numeric value or maximum string length.

Reply:
For the purposes of spec and ease of use we can: accept string representation of DRAND beacon directly. More specifically: maximum 64 characters, hex-encoded 32 bytes. I will also rename seed to beacon to avoid confusion with the rng seed in a general sense.

Diving deep into implementation: 
The RNG i intend to use is PCG, which is seed with 128-bits of internal state, typically supplied as 2 64-bit high/low integers.

I can convert the 32byte long DRAND beacon into two 64-bit unsigned integers using the least significant 64 bits of the supplied big integer.

Reply:
I have adjusted the spec to simply use "beacon". I have also expanded the section on determinism to concretely specify how the seed for PCG random number generator is derived from the 32 byte beacon.